crates/timers/src/future.rs

```diff
@@ -90,18 +85,25 @@ impl TimeoutFuture {
         let inner = JsFuture::from(promise);
         TimeoutFuture { id, inner }
     }
+
+    /// Project the pin onto the inner future.
+    fn project_inner(self: Pin<&mut Self>) -> Pin<&mut JsFuture> {
+        // Unsafe: The field `inner` must not be borrowed without structural pinning.
+        unsafe { self.map_unchecked_mut(|s| &mut s.inner) }
+    }
 }
 
 impl Future for TimeoutFuture {
-    type Item = ();
-    type Error = ();
+    type Output = ();
 
-    fn poll(&mut self) -> Poll<(), ()> {
-        match self.inner.poll() {
-            Ok(Async::Ready(_)) => Ok(Async::Ready(())),
-            Ok(Async::NotReady) => Ok(Async::NotReady),
-            // We only ever `resolve` the promise, never reject it.
-            Err(_) => wasm_bindgen::throw_str("unreachable"),
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        match self.project_inner().poll(cx) {
```

Comment:
Similar to above, this doesn't need `unsafe` code, it can just use `Pin::new(&mut self.inner).poll(cx)`

Also, this entire thing can actually be massively simplified by using `set_timeout` directly (not using `JsFuture`). But we can save that for another PR if you like.

Reply:
If we use `set_timeout`, should the timer start when the future is created (`TimeoutFuture::new`), or when it is actually `poll`ed?

Reply:
The current impl uses the `TimeoutFuture::new` tick to initiate the timeout. This results in a slightly more complex state machine.

Reply:
That's a really good question. I think there's reasonable arguments for both behaviors, but having it activate when polling is more logically consistent with `async`.

In any case we can discuss that in a new PR.